Index: linux-3.3.8/arch/mips/ath79/mach-ta361.c
===================================================================
--- linux-3.3.8.orig/arch/mips/ath79/mach-ta361.c
+++ linux-3.3.8/arch/mips/ath79/mach-ta361.c
@@ -22,6 +22,7 @@
 #include <linux/ar8216_platform.h>
 
 #include <asm/mach-ath79/ar71xx_regs.h>
+#include <linux/gpio.h>
 
 #include "common.h"
 #include "dev-eth.h"
@@ -54,6 +55,10 @@
 #define TA361_WMAC_CALDATA_OFFSET	0x1000
 
 #define TA361_GPIO_EXPANDER_GPIO_SPI_CS     16  /* EXPANDER GPIO SPI CS */ 
+#define TA361_GPIO_4P2V_EN                  11  //high is enable
+
+#define TA361_GPIO_4P2V_EN_VAL            1
+
 #if 0
 static struct gpio_led ta361_leds_gpio[] __initdata = {
 	{
@@ -119,6 +124,9 @@ static void __init ta361_setup(void)
 	ath79_gpio_output_select(TA361_GPIO_EXPANDER_GPIO_SPI_CS,
 			QCA953X_GPIO_OUT_MUX_SPI_CS1);
 			
+	ath79_gpio_direction_select(TA361_GPIO_4P2V_EN,true);
+    gpio_set_value(TA361_GPIO_4P2V_EN,TA361_GPIO_4P2V_EN_VAL);
+			
 	ath79_register_m25p80(NULL);
 
     /* There is not led on TA361-00 cpu gpio, all of led is on expander gpio */
